react/slides/react-router-4.markdown

```diff
@@ -0,0 +1,481 @@
+  }
+
+  return <div>TOP SECRET...</div>
+}
+
+...
+```
+
+--
+
+Import `Redirect` and `Switch` from `react-router-dom` package:
+
+```js
+import { Switch, Redirect } from 'react-router-dom'
+```
+
+<br />
+
+Configure redirecting from `/old-path` to `/new-path`:
+
+```js
+<Switch>
+  <Redirect from='/old-path' to='/new-path' />
+  <Route path='/new-path' component={...} />
+</Switch>
+```
+
+<br />
+
+Use `push` flag to create new entry in the `History`:
+
+```js
+<Redirect from='/old-path' to='/new-path'      /> // history after redirecting: /home -> /new-path
+<Redirect from='/old-path' to='/new-path' push /> // history after redirecting: /home -> /old-path -> /new-path
+```
+
+---
+
+Cool. But we need 404 fallback!
+
+<br />
+
+Import `Switch` from 'react-router-dom' package:
+
+```js
+import { Switch } from 'react-router-dom'
+```
+
+<br />
+
+And configure `Route` with fallback component:
+
+```js
+<Switch>
+  <Route to='/a' component={...} />
+  <Route to='/b' component={...} />
+  <Route to='/c' component={...} />
+
+  ...
+
+  // Notice that fallback route should be the last one
+  <Route component={() => '404 - Not Found'} />
+</Switch>
+```
+
+---
+
+Cool. But we want to control everything programmatically!
+
+Add `react-router-redux` package to project dependencies:
+
+<br />
+
+To install with `yarn`, run:
+
+```bash
+yarn add react-router-redux@next
+```
+
+<br />
+
+To install with `npm`, run:
+
+```bash
+npm install --save react-router-redux@next
+```
+
+--
+
+Notify `Redux` store about new reducer:
+
+<br />
+
+```js
+import { combineReducers } from 'redux'
+
+import { routerReducer } from 'react-router-redux'
+
+export default combineReducers({
+  ..., router: routerReducer
+})
+```
+
+--
+
+Notify `Redux` store about new middleware:
+
+```js
+import { routerMiddleware as createRouterMiddleware } from 'react-router-redux'
+
+...
+
+const configureStore = history => {
+  ...
+
+  const routerMiddleware =
+    createRouterMiddleware(
+      history
+    )
+
+  const middlewares = applyMiddleware(
+    ..., routerMiddleware, ...
+  )
+
+  ...
+}
+```
+
+--
+
+Update entry point of your application:
+
+```js
+import { createBrowserHistory } from 'history'
+import { ConnectedRouter } from 'react-router-redux'
+
+...
+
+import { configureStore } from './redux-store'
+
+const history = createHashHistory()
+const store = configureStore(history)
+
+...
+
+ReactDOM.render(
+  <Provider store={store}>
+    <ConnectedRouter history={history}>
+      <Application />
+    </ConnectedRouter>
+  </Provider>,
+
+  document.getElementById('app')
+)
+
+...
+```
+
+---
+
+Dispatch `push` and `replace` actions:
+
+```js
+import { push, replace } from 'react-router-redux'
+
+const Page = props => (
+  <div>
+    ...
+
+    // acts as Link
+    <button onClick={() => props.push('/abc')}>
+      Click me!
+    <button>
+
+    // acts as Redirect
+    <button onClick={() => props.replace('/abc')}>
+      Click me!
+    <button>
```

Comment:
not a nice example to use button, where we should use a Link

Reply:
Just replace with `<a>...</a>`?

Reply:
I wanted to show that we can also perform routing via dispatching Redux actions.

Reply:
while this works and is an example of showing how to use a 'push' and 'replace', I see no reason why we should use this in real world (we have Link for this purpose)